### ISSUE TITLE
issues with older xdg-desktop-portal versions

### ISSUE THREAD

**Issue by ppd** [ppd/freecad-ppd#15]
Older portal versions proxy everything through the document portal.

Problems:
- Files get lost when saving because FreeCAD appends a suffix, which is not allowed via the document portal
- Opening files with external links fail because the externally linked files are not available via the document portal

Possible solutions:
- simply do not use the native file dialog (ugly, can access fewer locations on the system)
- reject paths starting with `/run/user` and switch to the internal dialog at runtime. Would have to be patched upstream to be maintainable. Also an ugly solution.

**Comment by ppd** [comment]
Realthunder's solution:
https://github.com/realthunder/FreeCAD/commit/71adbe805a2315fc85f8e7fd80faf00386128765
https://github.com/realthunder/FreeCAD/commit/62d7d3f36d201c88faf2d6392d30f8f891217d09

@luzpaz Should we just switch to the internal file chooser by default? It's uglier, but it works for external links on all systems.

**Comment by ppd** [comment]
Idea: /etc/lsb-release is mounted from the host; we can parse it to get the host's OS & version. We could simply integrate a whitelist for known working hosts.

```python
whitelist = [
  "Ubuntu 21.10"
]

def get_os_info():
  os_info = {}
  with open("/etc/lsb-release") as f:
    for line in f:
      [key, value] = line.split("=")
      os_info[key] = value.strip("\"'\n ")
  return os_info

def is_in_whitelist(distrib_id, distrib_release):
  return f"{distrib_id} {distrib_release}" in whitelist

def host_supports_unproxied_file_chooser():
  os_info = get_os_info()
  return is_in_whitelist(
    os_info["DISTRIB_ID"],
    os_info["DISTRIB_RELEASE"]
  )

```

**Comment by ppd** [comment]
Bug filed for default filename support in Qt's QXdgDesktopPortalFileDialog: https://bugreports.qt.io/browse/QTBUG-100297

**Comment by luzpaz** [comment]
CC @hfiguiere ☝️  (related to Flatpak as well)

**Comment by ppd** [comment]
Support for unproxied access to files was introduced in xdg-desktop-portal v1.8.1: https://github.com/flatpak/xdg-desktop-portal/commit/48a981ee3cc3fbfbb2474629ab6c2e8a81ba67a4

**Comment by ppd** [comment]
Internal dialog by default was introduced in revisions:
- 139 for stable
- 140 for edge
- 
The native dialog can be forced by creating a boolean parameter `DontUseNativeDialog` = `false` in the parameter editor under `BaseApp/Preferences/Dialog` (see screenshot above)

**Comment by ppd** [comment]
I found another problem with the portal: While it works really well for a lot of things in the user's home folder, it does _not_ work properly for accessing things in `$SNAP`, i.e. in `/snap/freecad-ppd/current`. It kinda works, technically spoken, but not really from the average user's perspective. E.g. selecting patterns in TechDraw suffers from this.

**Comment by ToniTen** [comment]
Found yet another issue with portals in xubuntu 18.04 (Can't upgrade due to issues with GPU drivers, don't ask). Adding it here instead of creating a new issue. I will move this to a new issue if asked

I was trying to add a new directory for macro icons so I could have recognizable toolbar buttons for them, and the following happened:

On a pristine snap install, I go to Tools/Customize/Macros, click on the button next to Pixmap, click on the button "Icon folders...", and click on the button "+". A file chooser dialog opens, but it it doesn't let me select a directory, if I select a directory, it automatically opens it. If I click on an icon and select it, it proxies it through the document portal and I end with a path such as this "/run/user/1000/doc/97603f7e/dowel_pin_hole.svg". If I select the directory on the recents tab and click open, FreeCAD seems to lock up and I need to kill -9 it

The only way to get a custom directory for icons is to select a file in the file chooser, accept that it goes through the portal and gives me a path through /run/user/... and then open ~$/snap/freecad-ppd/common/user.cfg with your choice of text editor and replace the /run/user/... path with the correct path.

FreeCAD version info
```
OS: Ubuntu Core 20 (xubuntu)
Word size of FreeCAD: 64-bit
Version: 0.20.28730 (Git)
Build type: Release
Branch: master
Hash: 4358a9ba7031a61f1ade26958b74f95ff8daf1cf
Python 3.8.10, Qt 5.15.3, Coin 4.0.0, OCC 7.6.1
Locale: Spanish/Spain (es_ES)
```

**Comment by ppd** [comment]
Yeah, the portals are like a loosely coupled dependency that is very unstable because it exists in different implementations (KDE, Gnome...) and many versions. Significant development efforts need to be spent to get FreeCAD to play along 100% with confinement and portals. I wonder if simply polishing the internal file dialog would be a very good interim step. We could special-case running inside flatpaks/snaps and add proper shortcuts to all relevant paths, so users don't even see the strange home directory of the snap.

**Comment by ToniTen** [comment]
Doesn't change the fact that it's not possible to properly set a directory for custom icons, even if the directory is in say ~$/images/freecad_icons without manually editing the XML user.cfg file

EDIT: Oh, this also happens on RealThunder's LinkBranch3, but I thought it would be best to post it here first and then link it on LinkBranch3

**Comment by ppd** [comment]
But portals are not the default any more. Not in this snap and especially not in realthunder's snap. He explicitly disabled the portal file choosers, whereas I merely set the internal file chooser as the default.

Edit: Seems like this part of FreeCAD does not use FreeCAD's file dialog, but always defaults to the native dialog. That's certainly an interesting find. Thank you.

**Comment by ToniTen** [comment]
> But portals are not the default any more. Not in this snap and especially not in realthunder's snap. He explicitly disabled the portal file choosers, whereas I merely set the internal file chooser as the default.

Then why, on a pristine install of this snap, am I getting this?
![screenshot_2022-04-21_13-53-05](https://user-images.githubusercontent.com/55888212/164452656-fb438f94-45f7-4085-b076-cf3645637d7a.png)
![Screenshot_2022-04-21_13-51-26](https://user-images.githubusercontent.com/55888212/164452659-cd193df2-565c-45f4-9ee8-444a8c3e2a81.png)
